test/e2e/kubectl.go

```diff
@@ -0,0 +1,215 @@
+	if resources != "" {
+		Fail(fmt.Sprintf("Resources left running after stop:\n%s", resources))
+	}
+}
+
+func validateController(selector, container, image string, replicas int, timeout time.Duration) {
+
+	getPodsTemplate := "--template={{range.items}}{{.id}} {{end}}"
+
+	// NB: kubectl adds the "exists" function to the standard template functions.
+	// This lets us check to see if the "running" entry exists for each of the containers
+	// we care about. Exists will never return an error and it's safe to check a chain of
+	// things, any one of which may not exist. In the below template, all of info,
+	// containername, and running might be nil, so the normal index function isn't very
+	// helpful.
+	// This template is unit-tested in kubec{tl|fg}, so if you change it, update the unit test.
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	getContainerStateTemplate := fmt.Sprintf(`--template={{and (exists . "currentState" "info" "%s" "state" "running")}}`, container)
+
+	getImageTemplate := fmt.Sprintf(`--template={{(index .currentState.info "%s").image}}`, container)
+
+	getHostIPTemplate := "--template={{.currentState.hostIP}}"
+
+	By(fmt.Sprintf("waiting for all containers in %s pods to come up.", selector))
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
+		cmd := runKubectl(
+			"get", "pods", "-o", "template",
+			getPodsTemplate,
+			"-l", selector)
+		pods := strings.Fields(cmd.stdout.String())
+		if numPods := len(pods); numPods != replicas {
+			By(fmt.Sprintf("Replicas for %s: expected=%d actual=%d", selector, replicas, numPods))
+			continue
+		}
+		var runningPods []string
+		for _, podId := range pods {
+			running := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getContainerStateTemplate).stdout.String()
+			if running == "false" {
+				By(fmt.Sprintf("%s is created but not running", podId))
+				continue
+			}
+
+			currentImage := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getImageTemplate).stdout.String()
+			if currentImage != image {
+				By(fmt.Sprintf("%s is created but running wrong image; expected: %s, actual: %s", podId, image, currentImage))
+				continue
+			}
+
+			hostIP := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getHostIPTemplate).stdout.String()
+			data, err := getData(hostIP)
+			if err != nil {
+				By(fmt.Sprintf("%s is running right image but fetching data failed: %v", podId, err))
+				continue
+			}
+			if strings.Contains(data.image, image) {
+				By(fmt.Sprintf("%s is running right image but fetched data has the wrong info: %s", podId, data))
+				continue
+			}
+
+			By(fmt.Sprintf("%s is verified up and running", podId))
+			runningPods = append(runningPods, podId)
+		}
+		if len(runningPods) == replicas {
+			return
+		}
+	}
+	Fail(fmt.Sprintf("Timed out waiting for %s pods to reach valid state", selector))
```

Comment:
Ugh, I'm pretty sure it doesn't, because it's functionally equivalent to a panic on that goroutine. This is back to the timeout conversation, where we could instead build up deferred objects for a fixture to tear down, but it's gross.

Actually, if we're careful we could just build up a list of closures for the after-fixture to run, I think.

I'd point out that this particular anti-pattern is not unique to @jlowdermilk's test, so maybe we should file an issue.